runtime/compiler/runtime/J9CodeCacheManager.cpp

```diff
@@ -355,18 +359,46 @@ J9::CodeCacheManager::allocateCodeCacheSegment(size_t segmentSize,
 
    void *defaultEndAddress = vmemParams.endAddress;
 
+   uintptr_t someJitLibraryAddress = self()->getSomeJitLibraryAddress();
+
+#if defined(TR_TARGET_POWER)
+   size_t alignment = 64 * 1024;
+#elif (defined(LINUX) && defined(TR_TARGET_X86))
+   size_t alignment = 2 * 1024 * 1024;
+#elif (defined(TR_TARGET_S390))
+   size_t alignment = 1024 * 1024;
+#else
+   size_t alignment = pageSizes[0];
+#endif
+   if (largeCodePageSize > 0)
+      alignment = largeCodePageSize;
+
    if (preferredStartAddress)
       {
-   // J9PORT_VMEM_ADDRESS_HINT is only supported for default pages and linux
+      vmemParams.options |= J9PORT_VMEM_STRICT_ADDRESS; // if we cannot allocate a block in the desired range, return NULL
+      vmemParams.alignmentInBytes = alignment;
 #if defined(LINUX)
-      if (largeCodePageSize == 0)
+      vmemParams.options |= J9PORT_VMEM_ALLOC_QUICK; // Use smaps for guidance
+      // If jit library address is greater than MAX_DISTANCE (~2GB),
+      // then we search in the addresses before the jit library address.
+      // If jit library address is less than MAX_DISTANCE, (which means we
+      // do not have the full MAX_DISTANCE range before the jit library),
+      // then we search in the addresses after the jit library address
+      if (someJitLibraryAddress > MAX_DISTANCE_NEAR_JITLIBRARY_TO_AVOID_TRAMPOLINE)
          {
-         vmemParams.options |= J9PORT_VMEM_ADDRESS_HINT;
+         // Align the startAddress to page boundary
+         vmemParams.startAddress = (void *)OMR::align((size_t)(someJitLibraryAddress - MAX_DISTANCE_NEAR_JITLIBRARY_TO_AVOID_TRAMPOLINE), alignment);
+         vmemParams.endAddress = (void *)(((UDATA)preferredStartAddress) + SAFE_DISTANCE_REPOSITORY_JITLIBRARY);
```

Comment:
shouldn't this be
```
vmemParams.endAddress = (void *)(((UDATA)vmemParams.startAddress) + SAFE_DISTANCE_REPOSITORY_JITLIBRARY);
```
?

Reply:
When this function is called, `preferredAddress`  is set to `someJITDllAddress - SAFE_DISTANCE - repoSize`.  SAFE_DISTANCE is 64MB and accounts for the maximum size of the JIT dll plus some extra space.
Before my change we were initially searching the space from `preferredAddress` to `preferredStartAddress + SAFE_DISTANCE` by advancing the start point page by page. If this failed, we extended the search space to almost 2 GB, but used the smaps to guide the allocation.
Since lately the first search fails consistently, I basically removed it and used the smaps to search a space of 2GB-64MB in size, before the JIT dll.
I kept the endAddress like in the original version (preferredAddress+SAFE_DISTANCE). If we don't want to rely on `prefferedAddress` we could use endAddress = someJITDllAddr-repoSize.

Reply:
Ah I see; we should probably do something (not in this PR) to remove this implicit assumption about `preferredAddress`, since anyone reading just this method won't know that the math is based on what was passed in as an argument. Maybe for now just a comment will do?